File: client-android/app/src/test/java/stasis/test/client_android/mocks/MockServerApiEndpointClient.kt
package stasis.test.client_android.mocks

import okio.ByteString
import stasis.client_android.lib.api.clients.ServerApiEndpointClient
import stasis.client_android.lib.api.clients.exceptions.ResourceMissingFailure
import stasis.client_android.lib.model.DatasetMetadata
import stasis.client_android.lib.model.server.api.requests.CreateDatasetDefinition
import stasis.client_android.lib.model.server.api.requests.CreateDatasetEntry
import stasis.client_android.lib.model.server.api.requests.ResetUserPassword
import stasis.client_android.lib.model.server.api.requests.UpdateDatasetDefinition
import stasis.client_android.lib.model.server.api.responses.CreatedDatasetDefinition
import stasis.client_android.lib.model.server.api.responses.CreatedDatasetEntry
import stasis.client_android.lib.model.server.api.responses.Ping
import stasis.client_android.lib.model.server.api.responses.UpdatedUserSalt
import stasis.client_android.lib.model.server.datasets.DatasetDefinition
import stasis.client_android.lib.model.server.datasets.DatasetDefinitionId
import stasis.client_android.lib.model.server.datasets.DatasetEntry
import stasis.client_android.lib.model.server.datasets.DatasetEntryId
import stasis.client_android.lib.model.server.devices.Device
import stasis.client_android.lib.model.server.devices.DeviceId
import stasis.client_android.lib.model.server.schedules.Schedule
import stasis.client_android.lib.model.server.schedules.ScheduleId
import stasis.client_android.lib.model.server.users.User
import stasis.client_android.lib.utils.Try
import stasis.client_android.lib.utils.Try.Failure
import stasis.client_android.lib.utils.Try.Success
import java.time.Instant
import java.util.UUID
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicReference

open class MockServerApiEndpointClient(
    override val self: DeviceId
) : ServerApiEndpointClient {
    private val stats: Map<Statistic, AtomicInteger> = mapOf(
        Statistic.DatasetEntryCreated to AtomicInteger(0),
        Statistic.DatasetEntryDeleted to AtomicInteger(0),
        Statistic.DatasetEntryRetrieved to AtomicInteger(0),
        Statistic.DatasetEntryRetrievedLatest to AtomicInteger(0),
        Statistic.DatasetEntriesRetrieved to AtomicInteger(0),
        Statistic.DatasetDefinitionCreated to AtomicInteger(0),
        Statistic.DatasetDefinitionUpdated to AtomicInteger(0),
        Statistic.DatasetDefinitionDeleted to AtomicInteger(0),
        Statistic.DatasetDefinitionRetrieved to AtomicInteger(0),
        Statistic.DatasetDefinitionsRetrieved to AtomicInteger(0),
        Statistic.PublicSchedulesRetrieved to AtomicInteger(0),
        Statistic.PublicScheduleRetrieved to AtomicInteger(0),
        Statistic.DatasetMetadataWithEntryIdRetrieved to AtomicInteger(0),
        Statistic.DatasetMetadataWithEntryRetrieved to AtomicInteger(0),
        Statistic.UserRetrieved to AtomicInteger(0),
        Statistic.UserSaltReset to AtomicInteger(0),
        Statistic.UserPasswordUpdated to AtomicInteger(0),
        Statistic.DeviceRetrieved to AtomicInteger(0),
        Statistic.DeviceKeyPushed to AtomicInteger(0),
        Statistic.DeviceKeyPulled to AtomicInteger(0),
        Statistic.DeviceKeyExists to AtomicInteger(0),
        Statistic.Ping to AtomicInteger(0)
    )

    private val deviceSecretRef: AtomicReference<ByteString?> = AtomicReference(null)

    override val server: String = "mock-api-server"

    override suspend fun datasetDefinitions(): Try<List<DatasetDefinition>> {
        stats[Statistic.DatasetDefinitionsRetrieved]?.getAndIncrement()
        return Success(emptyList())
    }

    override suspend fun datasetDefinition(definition: DatasetDefinitionId): Try<DatasetDefinition> {
        stats[Statistic.DatasetDefinitionRetrieved]?.getAndIncrement()
        return Failure(RuntimeException("Test failure"))
    }

    override suspend fun createDatasetDefinition(request: CreateDatasetDefinition): Try<CreatedDatasetDefinition> {
        stats[Statistic.DatasetDefinitionCreated]?.getAndIncrement()
        return Success(CreatedDatasetDefinition(definition = UUID.randomUUID()))
    }

    override suspend fun updateDatasetDefinition(
        definition: DatasetDefinitionId,
        request: UpdateDatasetDefinition
    ): Try<Unit> {
        stats[Statistic.DatasetDefinitionUpdated]?.getAndIncrement()
        return Success(Unit)
    }

    override suspend fun deleteDatasetDefinition(definition: DatasetDefinitionId): Try<Unit> {
        stats[Statistic.DatasetDefinitionDeleted]?.getAndIncrement()
        return Success(Unit)
    }

    override suspend fun datasetEntries(definition: DatasetDefinitionId): Try<List<DatasetEntry>> {
        stats[Statistic.DatasetEntriesRetrieved]?.getAndIncrement()
        return Success(emptyList())
    }

    override suspend fun datasetEntry(entry: DatasetEntryId): Try<DatasetEntry> {
        stats[Statistic.DatasetEntryRetrieved]?.getAndIncrement()
        return Failure(RuntimeException("Test failure"))
    }

    override suspend fun latestEntry(
        definition: DatasetDefinitionId,
        until: Instant?
    ): Try<DatasetEntry?> {
        stats[Statistic.DatasetEntryRetrievedLatest]?.getAndIncrement()
        return Failure(RuntimeException("Test failure"))
    }

    override suspend fun createDatasetEntry(request: CreateDatasetEntry): Try<CreatedDatasetEntry> {
        stats[Statistic.DatasetEntryCreated]?.getAndIncrement()
        return Success(CreatedDatasetEntry(entry = UUID.randomUUID()))
    }

    override suspend fun deleteDatasetEntry(entry: DatasetEntryId): Try<Unit> {
        stats[Statistic.DatasetEntryDeleted]?.getAndIncrement()
        return Success(Unit)
    }

    override suspend fun publicSchedules(): Try<List<Schedule>> {
        stats[Statistic.PublicSchedulesRetrieved]?.getAndIncrement()
        return Success(emptyList())
    }

    override suspend fun publicSchedule(schedule: ScheduleId): Try<Schedule> {
        stats[Statistic.PublicScheduleRetrieved]?.getAndIncrement()
        return Failure(RuntimeException("Test failure"))
    }

    override suspend fun datasetMetadata(entry: DatasetEntryId): Try<DatasetMetadata> {
        stats[Statistic.DatasetMetadataWithEntryIdRetrieved]?.getAndIncrement()
        return Success(DatasetMetadata.empty())
    }

    override suspend fun datasetMetadata(entry: DatasetEntry): Try<DatasetMetadata> {
        stats[Statistic.DatasetMetadataWithEntryRetrieved]?.getAndIncrement()
        return Success(DatasetMetadata.empty())
    }

    override suspend fun user(): Try<User> {
        stats[Statistic.UserRetrieved]?.getAndIncrement()
        return Failure(RuntimeException("Test failure"))
    }

    override suspend fun resetUserSalt(): Try<UpdatedUserSalt> {
        stats[Statistic.UserSaltReset]?.getAndIncrement()
        return Success(UpdatedUserSalt("test-salt"))
    }

    override suspend fun resetUserPassword(request: ResetUserPassword): Try<Unit> {
        stats[Statistic.UserPasswordUpdated]?.getAndIncrement()
        return Success(Unit)
    }

    override suspend fun device(): Try<Device> {
        stats[Statistic.DeviceRetrieved]?.getAndIncrement()
        return Failure(RuntimeException("Test failure"))
    }

    override suspend fun pushDeviceKey(key: ByteString): Try<Unit> {
        stats[Statistic.DeviceKeyPushed]?.getAndIncrement()
        deviceSecretRef.set(key)
        return Success(Unit)
    }

    override suspend fun pullDeviceKey(): Try<ByteString> {
        stats[Statistic.DeviceKeyPulled]?.getAndIncrement()
        return when (val secret = deviceSecretRef.get()) {
            null -> Failure(ResourceMissingFailure())
            else -> Success(secret)
        }
    }

    override suspend fun deviceKeyExists(): Try<Boolean> {
        stats[Statistic.DeviceKeyExists]?.getAndIncrement()
        return Success(false)
    }

    override suspend fun ping(): Try<Ping> {
        stats[Statistic.Ping]?.getAndIncrement()
        return Success(Ping(id = UUID.randomUUID()))
    }

    val statistics: Map<Statistic, Int>
        get() = stats.mapValues { it.value.get() }

    val deviceSecret: ByteString?
        get() = deviceSecretRef.get()

    sealed class Statistic {
        object DatasetEntryCreated : Statistic()
        object DatasetEntryDeleted : Statistic()
        object DatasetEntryRetrieved : Statistic()
        object DatasetEntryRetrievedLatest : Statistic()
        object DatasetEntriesRetrieved : Statistic()
        object DatasetDefinitionCreated : Statistic()
        object DatasetDefinitionUpdated : Statistic()
        object DatasetDefinitionDeleted : Statistic()
        object DatasetDefinitionRetrieved : Statistic()
        object DatasetDefinitionsRetrieved : Statistic()
        object PublicSchedulesRetrieved : Statistic()
        object PublicScheduleRetrieved : Statistic()
        object DatasetMetadataWithEntryIdRetrieved : Statistic()
        object DatasetMetadataWithEntryRetrieved : Statistic()
        object UserRetrieved : Statistic()
        object UserSaltReset : Statistic()
        object UserPasswordUpdated : Statistic()
        object DeviceRetrieved : Statistic()
        object DeviceKeyPushed : Statistic()
        object DeviceKeyPulled : Statistic()
        object DeviceKeyExists : Statistic()
        object Ping : Statistic()
    }

    companion object {
        operator fun invoke(): MockServerApiEndpointClient =
            MockServerApiEndpointClient(self = UUID.randomUUID())
    }
}

File: client-ios/lib/Tests/StasisClientLibTests/Api/Clients/MockServerApiEndpointClient.swift
import Foundation
@testable import StasisClientLib

actor MockServerApiEndpointClient: ServerApiEndpointClient {
    nonisolated let selfDevice: DeviceId
    nonisolated let server: String = "mock-api-server"

    struct CallCounts: Sendable {
        var definitionsRetrieved = 0
        var definitionRetrieved = 0
        var definitionCreated = 0
        var definitionUpdated = 0
        var definitionDeleted = 0
        var entriesRetrieved = 0
        var entryRetrieved = 0
        var entryLatestRetrieved = 0
        var entryCreated = 0
        var entryDeleted = 0
        var publicSchedulesRetrieved = 0
        var publicScheduleRetrieved = 0
        var metadataWithIdRetrieved = 0
        var metadataWithEntryRetrieved = 0
        var userRetrieved = 0
        var userSaltReset = 0
        var userPasswordUpdated = 0
        var deviceRetrieved = 0
        var deviceKeyPushed = 0
        var deviceKeyPulled = 0
        var deviceKeyExistsChecked = 0
        var pinged = 0
        var commandsRetrieved = 0
        var analyticsEntriesSent = 0
    }

    private(set) var calls = CallCounts()
    private(set) var lastRequest: (any Sendable)?
    private var datasetMetadataOverrides: [DatasetEntryId: DatasetMetadata] = [:]
    private var datasetDefinitionsOverride: [DatasetDefinition]?
    private var latestEntryOverrides: [DatasetDefinitionId: DatasetEntryId?] = [:]
    private let pingDisabled: Bool
    private let commandsDisabled: Bool
    private let createDatasetEntryDelay: TimeInterval
    private var datasetDefinitionFailure: (any Error)?
    private var latestEntryFailure: (any Error)?
    private var datasetEntryFailure: (any Error)?

    func lastRequest<T: Sendable>(as type: T.Type = T.self) -> T? {
        lastRequest as? T
    }

    init(
        selfDevice: DeviceId = UUID(),
        pingDisabled: Bool = false,
        commandsDisabled: Bool = false,
        createDatasetEntryDelay: TimeInterval = 0
    ) {
        self.selfDevice = selfDevice
        self.pingDisabled = pingDisabled
        self.commandsDisabled = commandsDisabled
        self.createDatasetEntryDelay = createDatasetEntryDelay
    }

    func setDatasetDefinitionFailure(_ error: any Error) {
        datasetDefinitionFailure = error
    }

    func setLatestEntryFailure(_ error: any Error) {
        latestEntryFailure = error
    }

    func setDatasetEntryFailure(_ error: any Error) {
        datasetEntryFailure = error
    }

    func setDatasetMetadataOverride(_ entry: DatasetEntryId, _ metadata: DatasetMetadata) {
        datasetMetadataOverrides[entry] = metadata
    }

    func setDatasetDefinitionsOverride(_ definitions: [DatasetDefinition]) {
        datasetDefinitionsOverride = definitions
    }

    func setLatestEntryOverride(_ definition: DatasetDefinitionId, _ entry: DatasetEntryId?) {
        latestEntryOverrides[definition] = entry
    }

    func datasetDefinitions() async throws -> [DatasetDefinition] {
        calls.definitionsRetrieved += 1
        if let override = datasetDefinitionsOverride { return override }
        return [TestGenerators.definition(), TestGenerators.definition()]
    }

    func datasetDefinition(definition: DatasetDefinitionId) async throws -> DatasetDefinition {
        calls.definitionRetrieved += 1
        if let datasetDefinitionFailure {
            throw datasetDefinitionFailure
        }
        return TestGenerators.definition(id: definition)
    }

    func createDatasetDefinition(request: CreateDatasetDefinition) async throws -> CreatedDatasetDefinition {
        calls.definitionCreated += 1
        lastRequest = request
        return CreatedDatasetDefinition(definition: UUID())
    }

    func updateDatasetDefinition(definition: DatasetDefinitionId, request: UpdateDatasetDefinition) async throws {
        calls.definitionUpdated += 1
        lastRequest = request
    }

    func deleteDatasetDefinition(definition: DatasetDefinitionId) async throws {
        calls.definitionDeleted += 1
    }

    func datasetEntries(definition: DatasetDefinitionId) async throws -> [DatasetEntry] {
        calls.entriesRetrieved += 1
        return [
            TestGenerators.entry(definition: definition),
            TestGenerators.entry(definition: definition),
            TestGenerators.entry(definition: definition)
        ]
    }

    func datasetEntry(entry: DatasetEntryId) async throws -> DatasetEntry {
        calls.entryRetrieved += 1
        if let datasetEntryFailure {
            throw datasetEntryFailure
        }
        return TestGenerators.entry(id: entry)
    }

    func latestEntry(definition: DatasetDefinitionId, until: Date?) async throws -> DatasetEntry? {
        calls.entryLatestRetrieved += 1
        if let latestEntryFailure {
            throw latestEntryFailure
        }
        if let override = latestEntryOverrides[definition] {
            return override.map { TestGenerators.entry(id: $0, definition: definition) }
        }
        return TestGenerators.entry(definition: definition)
    }

    func createDatasetEntry(request: CreateDatasetEntry) async throws -> CreatedDatasetEntry {
        calls.entryCreated += 1
        lastRequest = request
        if createDatasetEntryDelay > 0 {
            try await Task.sleep(nanoseconds: UInt64(createDatasetEntryDelay * 1_000_000_000))
        }
        return CreatedDatasetEntry(entry: UUID())
    }

    func deleteDatasetEntry(entry: DatasetEntryId) async throws {
        calls.entryDeleted += 1
    }

    func publicSchedules() async throws -> [Schedule] {
        calls.publicSchedulesRetrieved += 1
        return [
            TestGenerators.schedule(isPublic: true),
            TestGenerators.schedule(isPublic: true),
            TestGenerators.schedule(isPublic: true)
        ]
    }

    func publicSchedule(schedule: ScheduleId) async throws -> Schedule {
        calls.publicScheduleRetrieved += 1
        return TestGenerators.schedule(id: schedule, isPublic: true)
    }

    func datasetMetadata(entry: DatasetEntryId) async throws -> DatasetMetadata {
        calls.metadataWithIdRetrieved += 1
        return datasetMetadataOverrides[entry] ?? TestGenerators.emptyDatasetMetadata
    }

    func datasetMetadata(entry: DatasetEntry) async throws -> DatasetMetadata {
        calls.metadataWithEntryRetrieved += 1
        return datasetMetadataOverrides[entry.id] ?? TestGenerators.emptyDatasetMetadata
    }

    func user() async throws -> User {
        calls.userRetrieved += 1
        return TestGenerators.user()
    }

    func resetUserSalt() async throws -> UpdatedUserSalt {
        calls.userSaltReset += 1
        return UpdatedUserSalt(salt: "test-salt")
    }

    func resetUserPassword(request: ResetUserPassword) async throws {
        calls.userPasswordUpdated += 1
        lastRequest = request
    }

    func device() async throws -> Device {
        calls.deviceRetrieved += 1
        return TestGenerators.device()
    }

    func pushDeviceKey(key: Data) async throws {
        calls.deviceKeyPushed += 1
    }

    func pullDeviceKey() async throws -> Data {
        calls.deviceKeyPulled += 1
        return Data("test-key".utf8)
    }

    func deviceKeyExists() async throws -> Bool {
        calls.deviceKeyExistsChecked += 1
        return false
    }

    func ping() async throws -> Ping {
        calls.pinged += 1
        if pingDisabled {
            throw EndpointFailure(message: "[pingDisabled] is set to [true]")
        }
        return Ping(id: UUID())
    }

    func commands(lastSequenceId: Int64?) async throws -> [CommandAsJson] {
        calls.commandsRetrieved += 1
        if commandsDisabled {
            throw EndpointFailure(message: "[commandsDisabled] is set to [true]")
        }
        let logoutNone = CommandAsJson.CommandParametersAsJson(logoutUser: nil)
        let logoutWithReason = CommandAsJson.CommandParametersAsJson(logoutUser: .init(reason: "test"))
        let all = [
            CommandAsJson(sequenceId: 1, source: "user", target: nil, parameters: logoutNone, created: Date()),
            CommandAsJson(
                sequenceId: 2,
                source: "service",
                target: selfDevice,
                parameters: logoutWithReason,
                created: Date()
            ),
            CommandAsJson(sequenceId: 3, source: "user", target: nil, parameters: logoutNone, created: Date())
        ]
        return all.filter { $0.sequenceId > (lastSequenceId ?? 0) }
    }

    func sendAnalyticsEntry(_ entry: AnalyticsEntry) async throws {
        calls.analyticsEntriesSent += 1
    }
}
